src/json.hpp.re2c

```diff
@@ -6906,17 +7178,40 @@ class basic_json
 
         @throw std::range_error if passed value is out of range
         */
-        long double get_number() const
+        void get_number(basic_json& result) const
         {
-            // conversion
             typename string_t::value_type* endptr;
             assert(m_start != nullptr);
-            const auto float_val = std::strtold(reinterpret_cast<typename string_t::const_pointer>(m_start),
-                                                &endptr);
+    
+            // Parse it as an integer
+            if(*reinterpret_cast<typename string_t::const_pointer>(m_start) != '-')  {
+                // Unsigned
+                result.m_value.number_unsigned = strtoull(reinterpret_cast<typename string_t::const_pointer>(m_start),&endptr,10);
+                result.m_type = value_t::number_unsigned;
+            }
+            else  {
+                // Signed
+                result.m_value.number_integer = strtoll(reinterpret_cast<typename string_t::const_pointer>(m_start),&endptr,10);
+                result.m_type = value_t::number_integer;
+            }
 
-            // return float_val if the whole number was translated and NAN
-            // otherwise
-            return (reinterpret_cast<lexer_char_t*>(endptr) == m_cursor) ? float_val : NAN;
+            // Parse it as a double
+            const auto float_val = strtold(reinterpret_cast<typename string_t::const_pointer>(m_start),&endptr);
+            long double int_part;
+            const auto frac_part = std::modf(float_val, &int_part);
+
+            // Test if the double or integer is a better representation
```

Comment:
Could you please add a comment to briefly describe the three parts of the disjunction?